src/Relude/List/Reexport.hs

```diff
@@ -8,18 +14,163 @@ License:    MIT
 Maintainer: Kowainik <xrom.xkov@gmail.com>
 
 Reexports most of the "Data.List" and "Data.List.NonEmpty".
+
+Note, that Relude reexports 'head', 'tail', 'init', 'last' from
+"Data,List.NonEmpty" instead of the "Data.List", so these functions are safe to
+use.
+
+`relude` also provides custom type error for better experience with transition
+from lists to 'NonEmpty' with those functions.
 -}
 
 module Relude.List.Reexport
        ( module Data.List
-       , module Data.List.NonEmpty
-       , module GHC.Exts
+         -- * NonEmpty List
+       , NonEmpty (..)
+       , nonEmpty
+       , head
+       , init
+       , last
+       , tail
+       , sortWith
        ) where
 
+import Data.Kind (Type)
 import Data.List (break, cycle, drop, dropWhile, filter, genericDrop, genericLength,
                   genericReplicate, genericSplitAt, genericTake, group, inits, intercalate,
                   intersperse, isPrefixOf, iterate, map, permutations, repeat, replicate, reverse,
                   scanl, scanr, sort, sortBy, sortOn, splitAt, subsequences, tails, take, takeWhile,
                   transpose, uncons, unfoldr, unzip, unzip3, zip, zip3, zipWith, (++))
-import Data.List.NonEmpty (NonEmpty (..), head, init, last, nonEmpty, tail)
-import GHC.Exts (sortWith)
+import Data.List.NonEmpty (NonEmpty (..), nonEmpty)
+import GHC.Exts (Constraint, sortWith)
+import GHC.TypeLits (ErrorMessage (..), Symbol, TypeError)
+
+import qualified Data.List.NonEmpty as NE (head, init, last, tail)
+
+
+-- $setup
+-- >>> import Relude
+
+
+-- | For tracking usage of ordinary list with @head@-like functions.
+type IsNonEmpty (f :: Type -> Type) (a :: Type) (fun :: Symbol) =
+    (f ~ NonEmpty, CheckNonEmpty f a fun)
```

Comment:
I agree here that we shouldn't reexport these type families. They are internal implementation details. And we don't want to spoil global namespace.